File: DSA-problems-class-FCUP/DAA 20-21 problems/Volume 1/DAA004 - Números arranjadinhos (Desafio) (DONE)/DAA004.cpp
#include <iostream>
using namespace std;

int sumDigits(long num){
  int sum = 0;
  while(num != 0){
	sum += num%10;
	num/=10;
  }
  return sum;
}
//15 9    20
int main(){
  int t;
  cin >> t;
  while(t--){
    long n,left,right, val;
    int k;
    cin >> n >> k;
    left = n++;
    right = left + 1000000000000000000;
    while(left <= right) {
		val = (left-right)/2 + right;
		if(sumDigits(val) < k)
			left = val;
		else if(sumDigits(val) > k)
			right = val;
		else if(sumDigits(val) == k){
			break;
		}	
	}
    cout << val << endl;
  }
  return 0;
}

File: DSA-problems-class-FCUP/DAA 20-21 problems/Volume 1/DAA004 - Números arranjadinhos (Desafio) (DONE)/DAA004.java
import java.util.Scanner;

class DAA004{
    static int sumDigits(long num){
	int sum = 0;
	while(num!=0){
	    sum+=num%10;
	    num/=10;
	}
	return sum;
    }

    static long increaseSum(long n, int k){
	int diff = k - sumDigits(n);
	long aux = n;
	long mult = 10;
	while(diff > 0){
	    long num = (aux%mult)/(mult/10); //gets each digit going from the right to the left
	    
	    long numToSum = 9 - num;
	    if(numToSum>diff) 
		numToSum = diff;
	    
	    aux+= numToSum*mult/10;
	    diff-= numToSum;
	    mult*=10;
	}
	return aux;
    }

    static long findNextNumber(long n, int k){
	if(k - sumDigits(n) > 0)
	    return increaseSum(n+1,k);
	else{
	    long aux;
	    long mult = 1;
	    aux = k==sumDigits(n)?n+1:n;
	    
	    while(sumDigits(aux) >= k){
		aux /= 10;
		mult *= 10;
		    
	    }
	    aux++;
	    return increaseSum(aux*mult, k);
	
	}
    }

    public static void main(String[] args){
	Scanner stdin = new Scanner(System.in);
	int t = stdin.nextInt();
	for(int i=0;i<t;i++){
	    long n;
	    int k;
	    n = stdin.nextLong();
	    k = stdin.nextInt();
	    System.out.println(findNextNumber(n, k));
	}
    }
}
